### PR TITLE
driver: set network.host entitlement by default for container drivers

### LINKED ISSUE #2255: [v0.13] allow network.host daemon entitlement by default in container drivers
### Description

Based on comments in https://github.com/moby/buildkit/pull/4524#issuecomment-1937917305 , there shouldn't be a need to set `--allowed-entitlemtns network.host` as buildkitd flags when builder runs in a container as the container network is already isolated. At least when `--driver-opt network` does not equal `host`.

### PR DESCRIPTION
**Pull request by crazy-max**
fixes #2255

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
This seems to be a potentially breaking change (from security perspective), and has to be documented?

File: builder/builder.go [hunk 1]
```diff
@@ -642,3 +640,37 @@ func validateBuildkitEndpoint(ep string) (string, error) {
...
+		for _, e := range allowInsecureEntitlements {
+			if e == "network.host" {
+				hasNetworkHostEntitlement = true
+				break
+			}
+		}
+
+		if !hasNetworkHostEntitlement {
+			// always set network.host entitlement as container network is
+			// isolated for docker-container and kubernetes drivers
```
Comment (AkihiroSuda) [inline builder/builder.go:670]: In the comment could you explain the purpose of setting the entitlement by default?

**Comment by tonistiigi** [comment]
> This seems to be a potentially breaking change (from security perspective), 

It is not. The default networking for build step containers if builder was in container was already host (meaning host inside the container, not host of machine) and will remain like this in v0.13. This was without setting any `--network=host`.

File: builder/builder.go [hunk 2]
```diff
@@ -642,3 +640,37 @@ func validateBuildkitEndpoint(ep string) (string, error) {
...
+
+		var hasNetworkHostEntitlement bool
+		for _, e := range allowInsecureEntitlements {
+			if e == "network.host" {
+				hasNetworkHostEntitlement = true
+				break
+			}
+		}
+
+		if !hasNetworkHostEntitlement {
```
Comment (tonistiigi) [inline builder/builder.go:668]: I think this condition should be "if no insecure entitlements set by user"

Comment (crazy-max) [reply builder/builder.go:668]: you mean we should check for any insecure entitlements? So if `security.insecure` or `network.host` or whatever is set for `--allow-insecure-entitlement` then we should skip setting `--allow-insecure-entitlement=network.host`?

Comment (tonistiigi) [reply builder/builder.go:668]: yes, just to leave a way for users to manually configure entitlements as well to be as they want them, without us overwriting

**Review by tonistiigi — APPROVED** [review]

**Pull request merged**